lib/workers/repository/update/pr/changelog/source-bitbucket.ts

```diff
@@ -0,0 +1,159 @@
+import URL from 'node:url';
+import { logger } from '../../../../../logger';
+import type { Release } from '../../../../../modules/datasource/types';
+import * as allVersioning from '../../../../../modules/versioning';
+import * as memCache from '../../../../../util/cache/memory';
+import * as packageCache from '../../../../../util/cache/package';
+import { regEx } from '../../../../../util/regex';
+import type { BranchUpgradeConfig } from '../../../../types';
+import { getTags } from './bitbucket';
+import { slugifyUrl } from './common';
+import { addReleaseNotes } from './release-notes';
+import { getInRangeReleases } from './releases';
+import type { ChangeLogRelease, ChangeLogResult } from './types';
+
+const cacheNamespace = 'changelog-bitbucket-release';
+
+function getCachedTags(
+  endpoint: string,
+  versionScheme: string,
+  repository: string
+): Promise<string[]> {
+  const cacheKey = `getTags-${endpoint}-${versionScheme}-${repository}`;
+  const cachedResult = memCache.get<Promise<string[]>>(cacheKey);
+  // istanbul ignore if
+  if (cachedResult !== undefined) {
+    return cachedResult;
+  }
+  const promisedRes = getTags(repository);
+  memCache.set(cacheKey, promisedRes);
+  return promisedRes;
+}
+
+export async function getChangeLogJSON(
+  config: BranchUpgradeConfig
+): Promise<ChangeLogResult | null> {
+  const versioning = config.versioning!;
+  const currentVersion = config.currentVersion!;
+  const newVersion = config.newVersion!;
+  const sourceUrl = config.sourceUrl!;
+  const packageName = config.packageName!;
+  const sourceDirectory = config.sourceDirectory!;
+
+  logger.trace('getChangeLogJSON for bitbucket');
+  const version = allVersioning.get(versioning);
+
+  const parsedUrl = URL.parse(sourceUrl);
+  const protocol = parsedUrl.protocol!;
+  const host = parsedUrl.host!;
+  const pathname = parsedUrl.pathname!;
+
+  logger.trace({ protocol, host, pathname }, 'Protocol, host, pathname');
+  const baseUrl = protocol.concat('//', host, '/');
+  const apiBaseUrl = protocol.concat('//', 'api.', host, '/');
+  const repository = pathname
+    .slice(1)
+    .replace(regEx(/\/$/), '')
+    .replace(regEx(/\.git$/), '');
+
+  if (repository.split('/').length < 2) {
+    logger.info({ sourceUrl }, 'Invalid bitbucket URL found');
+    return null;
+  }
+
+  const releases = config.releases ?? (await getInRangeReleases(config));
+  if (!releases?.length) {
+    logger.debug('No releases');
+    return null;
+  }
+
+  // This extra filter/sort should not be necessary, but better safe than sorry
+  const validReleases = [...releases]
+    .filter((release) => version.isVersion(release.version))
+    .sort((a, b) => version.sortVersions(a.version, b.version));
+
+  if (validReleases.length < 2) {
+    logger.debug('Not enough valid releases');
+    return null;
+  }
+
+  let tags: string[];
+
+  async function getRef(release: Release): Promise<string | null> {
```

Comment:
Please do not use local functions

Reply:
i have copied this pattern from the github and gitlab impls.  do you want these to deviate?

Reply:
Deviating is fine as long it adheres to newer best practices.